webapp/apps/core/views.py

```diff
@@ -0,0 +1,178 @@
+    close the job is to finishing, and then it will also display the
+    job results if the job is done. Finally, it will render a 'job failed'
+    page if the job has failed.
+
+    Cases:
+        case 1: result is ready and successful
+
+        case 2: model run failed
+
+        case 3: query results
+          case 3a: all jobs have completed
+          case 3b: not all jobs have completed
+    """
+
+    try:
+        url = OutputUrl.objects.get(pk=pk)
+    except OutputUrl.DoesNotExist:
+        raise Http404
+
+    model = url.unique_inputs
+
+    taxcalc_vers_disp = get_version(url, 'taxcalc_vers', TAXCALC_VERSION)
+    webapp_vers_disp = get_version(url, 'webapp_vers', WEBAPP_VERSION)
+
+    context_vers_disp = {'taxcalc_version': taxcalc_vers_disp,
+                         'webapp_version': webapp_vers_disp}
+    if model.tax_result:
+        # try to render table; if failure render not available page
+        try:
+            context = get_result_context(model, request, url)
+        except Exception as e:
+            print('Exception rendering pk', pk, e)
+            traceback.print_exc()
+            edit_href = '/taxbrain/edit/{}/?start_year={}'.format(
+                pk,
+                model.first_year or START_YEAR  # sometimes first_year is None
+            )
+            not_avail_context = dict(edit_href=edit_href,
+                                     **context_vers_disp)
+            return render(
+                request,
+                'taxbrain/not_avail.html',
+                not_avail_context)
+
+        context.update(context_vers_disp)
+        return render(request, 'taxbrain/results.html', context)
+    elif model.error_text:
+        return render(request, 'taxbrain/failed.html',
+                      {"error_msg": model.error_text.text})
+    else:
+       job_ids = model.job_ids
+
+        try:
+            jobs_ready = dropq_compute.results_ready(job_ids)
+        except JobFailError as jfe:
+            print(jfe)
+            return render_to_response('taxbrain/failed.html')
+        print(job_ids)
+        if any(j == 'FAIL' for j in jobs_ready):
+            failed_jobs = [sub_id for (sub_id, job_ready)
+                           in zip(job_ids, jobs_ready)
+                           if job_ready == 'FAIL']
+
+            # Just need the error message from one failed job
+            error_msgs = dropq_compute.get_results([failed_jobs[0]],
+                                                   job_failure=True)
+            if error_msgs:
+                error_msg = error_msgs[0]
+            else:
+                error_msg = "Error: stack trace for this error is unavailable"
+            val_err_idx = error_msg.rfind("Error")
+            error = ErrorMessageTaxCalculator()
+            error_contents = error_msg[val_err_idx:].replace(" ", "&nbsp;")
+            error.text = error_contents
+            error.save()
+            model.error_text = error
+            model.save()
```

Comment:
@lucassz what do you think about moving the error messages into the `CoreRun` table?